url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/episodes/managers/UpnpManager.java

```diff
@@ -0,0 +1,397 @@
+package stv6.episodes.managers;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.xerces.parsers.DOMParser;
+import org.w3c.dom.Document;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import stv6.episodes.BasicEpisode;
+import stv6.http.HttpRequestor;
+import stv6.http.request.Request;
+import stv6.http.request.Response;
+
+/**
+ * Theoretically generic UPNP manager, but specifically
+ *  designed to work with the PS3MediaServer UPNP server
+ *  
+ * @author dhleong
+ *
+ */
+public class UpnpManager extends AbstractManager {
+    
+    private static final List<Path> basePaths = new LinkedList<Path>();
+    static {
+        basePaths.add(new Path("0", ""));
+    }
+
+    private static final String CONTENT_PATH = "/upnp/control/content_directory";
+    private static final String BROWSE_ACTION = "urn:schemas-upnp-org:service:ContentDirectory:1#Search";
+    
+    private static final String CLASS_FOLDER = "object.container.storageFolder";
+    private static final String CLASS_VIDEO  = "object.item.videoItem";
+    
+    private static final String DEFAULT_PROFILE_FILENAME = "PMS.conf";
+    private static final String ENV_PROFILE_PATH = "PMS_PROFILE";
+    private static final String PROPERTY_PROFILE_PATH = "pms.profile.path";
+    /** TODO This is NOT necessarily constant, and we should add a setting */
+    private static final String PROFILE_DIRECTORY_NAME = "PMS";
+    
+    private static final Pattern RESULT_PATTERN = Pattern.compile("<Result>(.*)</Result>");
+    
+    private String broadIp;
+    private int upnpPort = -1;
+    private boolean isPms = false;
+    
+    private String localFoldersInfo;
+    private final List<Path> localFolders = new LinkedList<Path>();
+    
+    public UpnpManager(String broadcastingIp, int upnpPort) {
+        this.broadIp = broadcastingIp;
+        this.upnpPort = upnpPort;
+    }
+    
+    public UpnpManager(String broadcastIp, int upnpPort, String localFoldersInfo) {
+        this(broadcastIp, upnpPort);
+        
+        this.localFoldersInfo = localFoldersInfo;
+    }
+    
+    public UpnpManager(String broadcastIp, boolean isPms) {
+        this.broadIp = broadcastIp;
+        
+        if (!isPms) 
+            return;
+        
+        this.isPms = true;
+        
+        reload(); // make sure we can load everything
+    }
+
+    @Override
+    public void reload() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.